// bson/src/constants.ts
export const BSON_VERSION_SYMBOL = Symbol.for('@@mdb.bson.version');

export const BSON_INT32_MIN = -0x80000000;
export const BSON_INT64_MIN = -Math.pow(2, 63);

export const JS_INT_MAX = Math.pow(2, 53);

export const JS_INT_MIN = -Math.pow(2, 53);

export const BSON_DATA_STRING = 2;

export const BSON_DATA_ARRAY = 4;

export const BSON_DATA_UNDEFINED = 6;

export const BSON_DATA_BOOLEAN = 8;

export const BSON_DATA_NULL = 10;

export const BSON_DATA_DBPOINTER = 12;

export const BSON_DATA_SYMBOL = 14;

export const BSON_DATA_INT = 16;

export const BSON_DATA_LONG = 18;

export const BSON_DATA_MIN_KEY = 0xff;

export const BSON_BINARY_SUBTYPE_DEFAULT = 0;

export const BSON_BINARY_SUBTYPE_BYTE_ARRAY = 2;

export const BSON_BINARY_SUBTYPE_UUID_NEW = 4;

export const BSON_BINARY_SUBTYPE_ENCRYPTED = 6;

export const BSON_BINARY_SUBTYPE_SENSITIVE = 8;

export const BSONType = Object.freeze({
  double: 1,
  string: 2,
  object: 3,
  array: 4,
  binData: 5,
  undefined: 6,
  objectId: 7,
  bool: 8,
  date: 9,
  null: 10,
  regex: 11,
  dbPointer: 12,
  javascript: 13,
  symbol: 14,
  javascriptWithScope: 15,
  int: 16,
  timestamp: 17,
  long: 18,
  decimal: 19,
  minKey: -1,
  maxKey: 127
} as const);
